File: TASK08/parser/parser.cpp
#pragma once
#include <iostream>

#include "../instruction/instruction.cpp"
#include "../token/token.cpp"

#include <vector>
using namespace std;

string error_not_opened = "unexpected token ']', no open loop.";
string error_not_closed = "unexpected token EOF, loop not closed.";

class SyntaxError {
    Token tok;
    string errorMsg;

public:
    SyntaxError(int tok_s, int col, int line, string msg): tok(tok_s, col, line) {
        errorMsg = msg;
    }

    string errors() {
        return "parser: " + tok.ParsePos() + ": " + errorMsg;
    }
};

class Parser {
public:
    Token current;
    vector<Token> tokens;
    int index;

    Parser(vector<Token> tokens): current(0, 0, 0) {
        this->tokens = tokens;
        index = -1;
    }

    void next() {
        index++;
        current = tokens[index];
    }

    Chunk* program() {
        ChunkBuilder cb;
        vector<Token> loop_stack;

        while(true) {
            next();
            switch(current.type) {
            case Eof:
                break;
            case Plus:
                cb.changeValue(1);
            case Minus:
                cb.changeValue(-1);
            case LeftArrow:
                cb.changePointer(-1);
            case RightArrow:
                cb.changePointer(1);
            case Comma:
                cb.inputByte();
            case Period:
                cb.outputByte();
            case LeftBracket:
                loop_stack.emplace_back(current);
                cb.startLoop();
            case RightBracket:
                if(loop_stack.size() == 0) {
                    throw SyntaxError(current.type, current.pos.column, current.pos.line, error_not_opened).errors();
                }
                loop_stack.pop_back();
                cb.endLoop();
            default:
                throw "parser: invalid token from scanner";
            }
        }

        if(loop_stack.size() > 0) {
            throw SyntaxError(current.type, current.pos.column, current.pos.line, error_not_closed).errors();
        }

        return cb.finalize();
    }
};

File: TASK07/parser/parser.go
package parser

import (
	"fmt"

	"code.skyseaee.com/sonic-self/TASK07/instruction"
	"code.skyseaee.com/sonic-self/TASK07/token"
)

func Parse(tokens <-chan token.Token) (*instruction.Chunk, error) {
	p := parser{tokens: tokens}
	return p.program()
}

type parser struct {
	tokens  <-chan token.Token
	current token.Token
}

type SyntaxError struct {
	Token   token.Token
	message error
}

func (e SyntaxError) Error() string {
	return fmt.Sprintf("parser: %s: %v", e.Token.Position, e.message)
}

var (
	ErrNotOpened = fmt.Errorf("unexpected token ']', no open loop.")
	ErrNotClosed = fmt.Errorf("unexpected token EOF, loop not closed.")
)

func (p *parser) next() {
	p.current = <-p.tokens
}

func (p *parser) program() (*instruction.Chunk, error) {
	var c instruction.ChunkBuilder
	var stack []token.Token // loop stack

Parserloop:
	for {
		switch p.next(); p.current.Type {
		case token.Eof:
			break Parserloop
		case token.Plus:
			c.ChangeValue(1)
		case token.Minus:
			c.ChangeValue(-1)
		case token.LeftArrow:
			c.ChangePointer(-1)
		case token.RightArrow:
			c.ChangePointer(1)
		case token.Comma:
			c.InputByte()
		case token.Period:
			c.OutputByte()
		case token.LeftBracket:
			stack = append(stack, p.current)
			c.StartLoop()
		case token.RightBracket:
			if len(stack) == 0 {
				return nil, &SyntaxError{p.current, ErrNotOpened}
			}
			stack = stack[:len(stack)-1]
			c.EndLoop()

		default:
			panic("parser: invalid token from scanner")
		}
	}

	if len(stack) > 0 {
		return nil, &SyntaxError{stack[len(stack)-1], ErrNotClosed}
	}

	return c.Finalized(), nil
}
